CromwellRefdiskManifestCreator/src/main/java/org/broadinstitute/manifestcreator/CromwellRefdiskManifestCreatorApp.java

```diff
@@ -0,0 +1,151 @@
+
+    Path rootPath = rootDirectory.toPath();
+    Stack<File> stack = new Stack<>();
+    stack.push(rootDirectory);
+    List<ReferenceFile> referenceFiles = Collections.synchronizedList(new ArrayList<>());
+    for (int i = 0; i < nThreads; i++) {
+      executorService.submit(() -> {
+        try {
+          doWork(rootPath, referenceFiles, stack, totalNumberOfFiles, countDownLatch);
+        } catch (CRC32Exception e) {
+          logger.error("Error occurred. Shutting down.", e);
+          executorService.shutdownNow();
+          System.exit(1);
+        }
+      });
+    }
+
+    countDownLatch.await();
+
+    String manifestFilePath = args[3];
+    File manifestFile = new File(manifestFilePath);
+    if (manifestFile.exists() || !manifestFile.createNewFile()) {
+      logger.error("File " + manifestFile.getAbsolutePath() + " already exists or cannot be created.");
+    } else {
+      referenceFiles.forEach(referenceFile -> manifest.getFiles().add(referenceFile));
+      new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(manifestFile, manifest);
+      logger.info("Finished. Total number of reference files: " + manifest.getFiles().size());
+    }
+
+    executorService.shutdown();
+  }
+
+  private static void doWork(Path rootPath, List<ReferenceFile> accumulator, Stack<File> stack, int totalNumberOfFiles, CountDownLatch countDownLatch) throws CRC32Exception {
+    while(!stack.isEmpty() || accumulator.size() != totalNumberOfFiles) {
+      if (stack.isEmpty()) {
+        continue;
+      }
+
+      File curFile = stack.pop();
+
+      if (curFile.isDirectory()) {
+        File[] children = curFile.listFiles();
+        if (children != null) {
+          Stream.of(children).forEach(stack::push);
+        }
+      } else {
+        ReferenceFile refFile = new ReferenceFile();
+
+        String relativePath = rootPath.relativize(curFile.toPath()).toString();
+        refFile.setPath(relativePath);
+
+        long crc32 = calculateCrc32(curFile);
+        refFile.setCrc32(crc32);
+
+        accumulator.add(refFile);
+        logger.info(Thread.currentThread().getName() + " finished processing file " + curFile.getAbsolutePath());
+      }
+    }
+
+    logger.info("Thread " + Thread.currentThread().getName() + " finished processing.");
+    countDownLatch.countDown();
+  }
+
+  private static long calculateCrc32(File file) throws CRC32Exception {
+    CRC32 crc32 = new CRC32();
+    int bufferSize = 32768;
+    try (InputStream is = new BufferedInputStream(new FileInputStream(file), bufferSize)) {
+      byte[] buff = new byte[bufferSize];
+      int bytesRead;
+      while ((bytesRead = is.read(buff)) != -1) {
+        crc32.update(buff, 0, bytesRead);
+      }
+    } catch (IOException e) {
+      throw new CRC32Exception("Cannot read from file " + file.getAbsolutePath(), e);
+    }
+    return crc32.getValue();
+  }
+
+  private static int countFiles(File directory) {
```

Comment:
Kind of surprising something like this isn't in the standard library. I also found this https://stackoverflow.com/a/41449174